src/util/route.js

```diff
@@ -6,13 +6,13 @@ export function isSameRoute (a: Route, b: ?Route): boolean {
   } else if (a.path && b.path) {
     return (
       a.path === b.path &&
-      a.hash === b.hash &&
+      (a.hash || '') === (b.hash || '') &&
```

Comment:
To sum up, @yyx990803 A more elegant fix for this (instead of the `hash || ''` fix) involves modifying `Location` type, or a converter from `Location` to `Route`:

1. `compareTarget` can be a `Location`
https://github.com/vuejs/vue-router/blob/next/src/components/link.js#L31

2. and it's passed into `isSameRoute` whose second param is type `?Route` (which means `Route | void | null` if I understood correctly), but flow didn't give a warning
https://github.com/vuejs/vue-router/blob/next/src/components/link.js#L33